lib/blemulator/internal/simulation_manager.dart

```diff
@@ -24,17 +24,92 @@ class SimulationManager {
 
   Future<void> _startDeviceScan() async {
     _peripherals.forEach((peripheral) {
-      _scanSubscriptions.add(peripheral
-          .onScan(allowDuplicates: true)
-          .listen((scanResult) {
-            print(scanResult);
-            return _bridge.addScanResult(scanResult);
-          }));
+      _scanSubscriptions
+          .add(peripheral.onScan(allowDuplicates: true).listen((scanResult) {
+        print(scanResult);
+        return _bridge.publishScanResult(scanResult);
+      }));
     });
   }
 
   Future<void> _stopDeviceScan() async {
     _scanSubscriptions.forEach((subscription) => subscription.cancel());
     _scanSubscriptions.clear();
   }
+
+  Future<SimulatedCharacteristic> _readCharacteristicForIdentifier(
+    int characteristicIdentifier,
+  ) {
+    SimulatedCharacteristic targetCharacteristic;
+    peripheralsLoop:
+    for (SimulatedPeripheral peripheral in _peripherals) {
+      for (SimulatedService service in peripheral.services()) {
+        SimulatedCharacteristic characteristic =
+            service.characteristic(characteristicIdentifier);
+        if (characteristic != null) {
+          targetCharacteristic = characteristic;
+          break peripheralsLoop;
+        }
+      }
+    }
+
+    if (targetCharacteristic == null)
+      return Future.error("Characteristic not found");
+
+    return targetCharacteristic.read().then(
+          (value) => targetCharacteristic..value = value,
+        );
+  }
+
+  Future<SimulatedCharacteristic> _readCharacteristicForDevice(
+    String peripheralId,
+    String serviceUuid,
+    String characteristicUUID,
+  ) {
+    SimulatedPeripheral targetPeripheral =
+        _peripherals.firstWhere((peripheral) => peripheral.id == peripheralId);
+
+    SimulatedCharacteristic targetCharacteristic = targetPeripheral
+        .services()
+        .firstWhere(
+          (service) => service.uuid == serviceUuid,
+          orElse: () => null,
+        )
+        ?.characteristics()
+        ?.firstWhere(
+          (characteristic) => characteristic.uuid == characteristicUUID,
+          orElse: () => null,
+        );
+
+    if (targetCharacteristic == null)
+      return Future.error("Characteristic not found");
+
+    return targetCharacteristic.read().then(
```

Comment:
```
await targetCharacteristic.read();
return targetCharacteristic;
```
or
```
SimulatedCharacteristic SimulatedCharacteristic.read();
```
?

Reply:
Actually, it should be `SimulatedCharacteristic._value` and since that should be private we should have some intermediate representation, like `CharacteristicResponse(Characteristic, Uint8List)`.